### PR TITLE
Do a general code review of the code

### PR DESCRIPTION
**Pull request by gglanzani**
I've added some TODOs which might not be need to stay there

### CODE REVIEW

File: evol/helpers/_utils.py [hunk 1]
```diff
@@ -10,6 +10,6 @@ def select_partition(length: int, min_size: int=1, max_size: int=None) -> Tuple[
     :param max_size: Maximum length of the partition. Defaults to length - 1.
     :return: Start and end index of the partition.
     """
-    partition_size = randint(min_size, length-1 if max_size is None else max_size)
+    partition_size = randint(min_size, max_size or length - 1)
```
Comment (koaning) [inline evol/helpers/_utils.py]: Although this is less characters, I find it less obvious that this does what it is supposed to do when reading it. What is the main reason for making this change?

File: evol/helpers/utils.py [hunk 2]
```diff
@@ -24,10 +24,7 @@ def offspring_generator(parents: List[Individual],
     while True:
         # Obtain parent chromosomes
         parents = parent_picker(parents, **kwargs)
-        if isinstance(parents, Individual):
-            chromosomes = (parents.chromosome,)
-        else:
-            chromosomes = tuple(individual.chromosome for individual in parents)
+        chromosomes = tuple(individual.chromosome for individual in parents)
```
Comment (koaning) [inline evol/helpers/utils.py]: +1

File: evol/helpers/utils.py [hunk 3]
```diff
@@ -53,3 +50,21 @@ def result(*args, **kwargs):
         except TypeError:
             return func(*args, **{k: v for k, v in kwargs.items() if k in signature(func).parameters})
     return result
+
+
+def ensure_list(func: Callable) -> Callable:
```
Comment (koaning) [inline evol/helpers/utils.py]: +1

File: evol/individual.py [hunk 4]
```diff
@@ -32,17 +33,17 @@ def evaluate(self, eval_function, lazy=False):
...
+        :type mutate_func: Callable[chromosome, **kwargs] -> chromosome
         :param probability: Probability that the individual mutates.
             The function is only applied in the given fraction of cases.
             Defaults to 1.0.
         :type probability: float
         :param kwargs: Arguments to pass to the mutation function.
         """
-        if probability == 1.0 or random() < probability:
-            self.chromosome = func(self.chromosome, **kwargs)
+        if isclose(probability, 1.0) or (random() < probability):
```
Comment (koaning) [inline evol/individual.py]: +1

File: evol/population.py [hunk 5]
```diff
@@ -10,13 +10,14 @@
...
 from evol import Individual
-from evol.helpers.utils import select_arguments, offspring_generator
+from evol.helpers.utils import select_arguments, offspring_generator, ensure_list
 
 
 class Population:
     """Population of Individuals
 
     :param chromosomes: Collection of initial chromosomes of the Population.
+    #TODO What is this Collection? Should it be a typing.List, or perhaps a typing.Sequence?
```
Comment (koaning) [inline evol/population.py]: I think `typing.Sequence` should be fine. I can't think of a reason why you couldn't give us a large tuple. @rogiervandergeer ?

File: evol/population.py [hunk 6]
```diff
@@ -151,16 +152,19 @@ def survive(self, fraction=None, n=None, luck=False) -> 'Population':
         :return: self
         """
         if fraction is None:
-            if n is None:
-                raise ValueError('everyone survives! must provide either "fraction" and/or "n".')
+            if (n is None) or (n > len(self.individuals)):
```
Comment (koaning) [inline evol/population.py]: In the future we'll have populations that can have islands which will allow for populations that are larger than the intended size. I do agree that if `n>intended_size` that we should send some form of warning.

File: evol/population.py [hunk 7]
```diff
@@ -188,24 +192,26 @@ def breed(self, parent_picker, combiner, population_size=None, **kwargs) -> 'Pop
             Arguments are only passed to the functions if they accept them.
         :return: self
         """
-        parent_picker = select_arguments(parent_picker)
+        # TODO I'm not exactly sure, on an API level, where the `ensure_list` and `select_arguments`
+        # should go, if here on inside offspring_generator
+        parent_picker = ensure_list(select_arguments(parent_picker))
```
Comment (koaning) [inline evol/population.py]: The parent pick is also allowed to return just 1 individual. In fact, technically, we even allow for no parent to be passed by the parent picker.

**Comment by gglanzani** [comment]
One extra thing: the docs should *not* be in git :)

File: evol/evolution.py [hunk 8]
```diff
@@ -140,29 +141,30 @@ def breed(self, parent_picker, combiner, population_size=None, name=None, **kwar
...
         :type probability: float
-        :param name: Name of the mutate step.
+        :param name: Name of the mutate_with step.
         :type name: str
         :param kwargs: Kwargs to pass to the parent_picker and combiner.
             Arguments are only passed to the functions if they accept them.
         :return: self
         """
-        return self._add_step(MutateStep(name=name, probability=probability, func=func, **kwargs))
+        return self._add_step(MutateStep(name=name, probability=probability, mutate_func=func, **kwargs))
```
Comment (rogiervandergeer) [inline evol/evolution.py]: The naming of the _function_ arguments is erratic in general. I think we should spend some time getting this right.

File: evol/helpers/_utils.py (hunk 1 above)
Comment (rogiervandergeer) [reply evol/helpers/_utils.py]: I do like it; it's much more compact than the original, although less explicit.

File: evol/population.py (hunk 5 above)
Comment (rogiervandergeer) [reply evol/population.py]: Yes, should be `Sequence`

File: evol/population.py (hunk 6 above)
Comment (rogiervandergeer) [reply evol/population.py]: This was intentional: the size of a population should be rather well defined when going through an evolution, but the size of an incoming population is completely free. I do not want to raise any exceptions when someone passes a smaller population than expected. If however `n > self.intended_size` we may want to raise.

File: evol/population.py (hunk 7 above)
Comment (rogiervandergeer) [reply evol/population.py]: I'm indeed not sure if we want to enforce the return type of the parent picker. I can live with not allowing a single individual but enforcing that to be a list (tuple?) with a single element. Can we think of use cases to return something else than an ordered collection of individuals? Something with weights?

File: examples/example_rock_paper_scissors.py [hunk 9]
```diff
@@ -60,7 +60,7 @@ def run_rock_paper_scissors(population_size=100, n_iterations=200, random_seed=4
         Evolution()
         .survive(fraction=survive_fraction)
         .breed(parent_picker=pick_random, combiner=lambda x, y: x.combine(y), n_parents=2)
-        .mutate(lambda x: x.mutate())
+        .mutate(lambda x: x.mutate_with())
```
Comment (rogiervandergeer) [inline examples/example_rock_paper_scissors.py]: This is the method of the `RockPaperScissorsPlayer`. If we decide to rename `mutate` to `mutate_with`, this must still be `mutate`.

**Comment by koaning** [comment]
@gglanzani whats against having the docs in git? i like having everything in one place.

**Comment by gglanzani** [comment]
@koaning They are generated, so it's duplicate. 

If you have them committed at least have a pre-commit hook to generate them again on code changes, and document how to create the hook (and possibly how to use `pdoc` which is not even in the `setup.py`).

See [here](https://stackoverflow.com/questions/4983060/should-generated-documentation-go-into-source-control) for some extra rationale on this.

**Comment by koaning** [comment]
That actually makes sense.

**Comment by gglanzani** [comment]
I think you need to add it locally (it's a pre-commit, not post-commit) and document it here.

(see https://github.com/git/git/blob/master/templates/hooks--pre-commit.sample for and example of pre-commit hooks).

**Comment by koaning** [comment]
this review was much appreciated but it was done so long ago that the merge would be horrendous. i am closing this to keep around for historical reasons but we won't merge it directly anymore.

**Pull request closed without merging**